### ISSUE TITLE
Dev Env: Instructions for using IntelliJ as our development IDE for Java and Javascript

### ISSUE THREAD

**Issue by bruderol** [scenarioo/scenarioo#185]
Adjust our Dev Env instructions to use intelliJ as our preferred IDE to contribute to scenarioo.
Remove all outdated and not anymore supported docu (like VM, eclipse, etc.)

**Acceptance Criteria**
* [x] IntelliJ Setup has been documented
* [x] Devs are working with this setup
* [ ] docu for the dev env has been cleaned up and consolidated as follows:
           - consolidate parts that are currently duplicated
           - remove old docu parts not needed anymore
           - such that we have on only truth and this is the IntelliJ setup, explained in such a way, that you are free to use other IDE, but mostly targeted to devs using IntelliJ which is our primary choice

* [ ] open other issues if there is some further left overs (???)

**Out of Scope:**
* For the code formatting topic I opened already a follow up issue, see #655

**Comment by bruderol** [comment]
@adiherzog @danielsuter 
I make my self an easter present today and start to setup my scenarioo development environment with pure IntelliJ setup. If somebody of you has allready experience with this, it would be great if you could share it with me here.

@danielsuter How are you currently developing? Pure IntelliJ too?

BTW: We allready have some open source licenses for intelliJ, in case somebody needs one.

**Comment by bruderol** [comment]
I started to describe the setup here:
https://github.com/scenarioo/scenarioo/wiki/Development-Environment-with-IntelliJ

**Comment by bruderol** [comment]
I also changed the existing setup pages in the WIKI refering to the four (!!!) different setup variants that we have currently descirbed up to now. We need to reduce this to ONE setup instruction (for setting up one environment)!

https://github.com/scenarioo/scenarioo/wiki/Development-Environment

**Comment by bruderol** [comment]
Moved the new dev env setup to be documented in sources directly:
https://github.com/scenarioo/scenarioo/blob/develop/developer.readme.md

**Comment by bruderol** [comment]
found out that we allready use .editorconfig for some formatting in the client, that seems to work well with IntelliJ, but there are some issues with formatting to be considered probably.

**Comment by bruderol** [comment]
Probleme mit gulp, protractor und gulp-protractor in windows:

1. protractor auf windows hat ein Problem mit optional dependencies (python, visual studio, ...) was zu komischen error messages bei npm install führt. Kann man aber vemrutlich ignorieren, unklar ob danach alles weitere noch funktioniert. (bei mir im Projekt war das ein Grund, protractor global zu installieren, wir haben keinen anderen Workaround gefunden, der bei allen Entwicklern funktioniert hat)

2. Folgende Zeile (9) im package.json scheint bei gradle clean install zu Problemen zu führen:
"postinstall": "./node_modules/gulp-protractor/node_modules/protractor/bin/webdriver-manager update", ohne "./..."-Pfad funktionert es bei mir erst mal mit dem builden ... stellt sich die Frage ob auch später noch alles tut ... und v.a. ob es auf dem CI oder der VM diesen Pfad unbedingt braucht?

Danach Probleme mit bower:

:scenarioo-client:bower FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':scenarioo-client:bower'.
> A problem occurred starting process 'command 'bower''
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
BUILD FAILED

**Comment by bruderol** [comment]
Ich kann zwar meinen gradle build nicht aus der IDE ankicken, siehe Probleme oben, folgendes scheint aber trotzdem zu funktionieren:

* run all tests to generate dummy example data on project 'scenarioo-docu-generation-example'
* Server starten (Run config Tomcat 7 inklusive Deployment des Build Artifacts)
* Client server von kommandozeile starten
* Scenarioo Webapp browsen

noch offen:
* Debuging von Java Server Code?
* e2e-Tests?

**Comment by bruderol** [comment]
except for some open untest points (see readme), I can confirm, that I can develop using my IntelliJ setup. 

I will now collect some experience with this setup. 

But I think it is very easy to setup, as you can see in my readme that covers everything you need to get started.

**Comment by danielsuter** [comment]
@bruderol I'm also developing with InteilliJ. But still figuring out what's the best way to set it up. Currently I'm importing scenarioo as a gradle project. Unfortunately I had to refresh it, so it would recognize gradle artefacts.
I managed debugging of client test code with intellij. I will figure out how to debug server code also. If I find a reasonable solution, I will document it.

**Comment by bruderol** [comment]
I can allready debug server code. allmost everything works with allmost no problems in my setup, except for some gradle tasks.

**Comment by danielsuter** [comment]
I only found a different page in our wiki which looked very basic. So I did it on my own. I will give you feedback on this one.

**Comment by bruderol** [comment]
@danielsuter Wow, another wiki page about dev env ... I cleaned those (4 other pages!) in the wiki up, but I did not see that one :wink: 

Ok, I removed this old page now.

This is the major dev env page in the Wiki:
https://github.com/scenarioo/scenarioo/wiki/Development-Environment

**Comment by bruderol** [comment]
I continued testing my setup:

Seems to work quite well so far. only small issues, mainly with gradle build and all this js-tools integration in gradle not working properly.

What works:

[x] run all java unit tests (from intellij)
[x] run and debug the server (from intellij)
[x] developing server and client code
[x] gulp build
[x] gulp test
[x] gulp test-e2e
[x] gulp serve

For me that is all I currently need.

What does not work yet (mainly because of windows and global dependencies installed on my machine, that I need):
[ ] some problem with gradle build because of failing optional protractor dependencies (when doing npm install, which fails the full gradle build)
[ ] some problems with updating the webdriver for gulp-protractor ... and unfortunately I can not use the task "webriver_update" because jslint does not like properties of the form "webdriver_update" in our code :-( [that is why I could not yet commit that, but I have a workaround for me localy]

What is open (needs further clarification):
[ ] code formatting and other settings need some validation, but since I never format something automatically (I do not like that anyway, to ensure that I not have changes I not realy wanted to change), I do not see why this could cause trouble.

What could be improved:
[ ] Provide some shared run configs for everybody to work more fast from intellij for most important tasks to perform when developing (in my current project we found a way to share this - and I recommend that, but we might have to discuss it - maybe something fro the coding days to optimize for all of us??).

@adiherzog @danielsuter What do you think, shouldnt we switch to that much better setup soon for most developers? I see many advantages of the new setup. E.g. the e2e tests seem to run much more stable and fast on my windows dev setup without vm (never before I could develop scenarioo so fast).

**Comment by danielsuter** [comment]
@bruderol 
I don't like the idea of only supporting IntelliJ. I find it a good idea to document it as well but every developer must be able to choose his or her IDE without too much trouble. I also prefer developing in IntelliJ but simply because I could get rid of the VM.

I personally like formatting all source code in the same style to make it more consistent. It's annoying if I format a file and have changes everywhere, simply because the author did not use proper formating. I usually make another commit just formating the file and then commit my changes. I know this makes merging really hard but the advantages of a consistent formatting are more important in my opinion.

What kind of problems do you have with your protractor tests. They're green on my machine :D. I also can easily update my webdriver. Also npm install works without troubles for me. Are you developing on develop with the lastest gradle build?

**Comment by bruderol** [comment]
Not agree on the formating part. when you develop on an open source project you should always take care to only format the parts that you realy changed! for reformatting code, we need to plan this to not cause troubles.

Never forget: right now, we have two active forks working on new features in parallel!

And there will allways be somebody fogeting to format his code or not using the right tooling for fomating. So the rule is: Do not format code that you did not have to touch!

**Comment by bruderol** [comment]
furhtermore I still think, that in the future we should all switch to intelliJ and have this setup as the preferred scenarioo dev setup (on VM as well as standalone). Because using the same dev env has many advantages: e.g.sharing run configs, sharing templates, sharing other stuff to make developer life more easier.

**Comment by bruderol** [comment]
@danielsuter @adiherzog Would be great if we could have a short look at the state of this in camp and decide whether we shouldn't recommend to use this setup in future. I think there are allmost no problems anymore with using this in windows (only one last problem with gradle build, that might be easy to solve). I am very happy with working with IntelliJ on my machine. And I am much more productive with this setup, than I was before with the slow VM (ok maybe it was just my slow old computer, but anyway).

**Comment by danielsuter** [comment]
I'm also happily working with IntelliJ and would recommend this setup.

**Comment by bruderol** [comment]
there are currently some issues with formattings (CRLF) on dola's and my setup ... need to see about that ... 

but its getting better every day ... update to IntelliJ 2016.1 and newest Markdown plugins is also reocmmended (i am using that) and I like the new JavaFX based Markdown viewer (even displays images inline) soooooo much :+1: :smile:

**Comment by bruderol** [comment]
Problem concerning CRLF and LF:
I think using following git settings that shouldnt be any problem anymore:
https://help.github.com/articles/dealing-with-line-endings/

For our project (since we should not have any windows specific files),
setting following should work:
```
git config core.autocrlf input
```

**Comment by bruderol** [comment]
Also i found out, that the gradle problem I has once, is gone, when switching the JVM for gradle to JVM 1.7 - maybe we have to switch our project properly to use JDK 1.8 for not having this issue anymore.

I will document that.

**Comment by bruderol** [comment]
ok, fixed and documented some detail issues. The setup becomes much better and better.

I do not see any reason any more to not use this setup.
(maybe except for the documented "untested" features)

**Comment by danielsuter** [comment]
I can even debug protractor and Karma tests :-)

**Comment by bruderol** [comment]
Cool :-) 👍 
I really think the setup is now allmost bullet proof - and if there are still problems we can still fix them.
I recommend that we check that we can also do the full release with that setup without problems.

One issue to solve might be the formatting (intelliJ removes trailing spaces also in empty Javadoc Comment lines and empty lines, where current eclipse settings dont):
I checked about that. IntelliJ takes the .editorconfig file into account that we have in our scenarioo repo. unfortuately our code is somehow formatted using slightly different format settings (probably because Eclipse ignores this file ? ) as documented here: https://github.com/scenarioo/scenarioo/wiki/Eclipse-IDE-Settings

Maybe we can align this. Not sure about that.

Proposal: 
* We format all the code to the new IntelliJ/Editorconfig formattings
* We document how this same settings can be used in Eclipse (if possible at all)
* Everybody that still uses Eclipse has to ensure to change those settings.

In general everybody should make sure anyway to not commit reformattings in line of codes that he did not really change and only format code that he really changed (in intelliJ this can be done with Ctrl+Alt+Shift+L and check (x) `only VCS changed text`)

One last remark concerning "final" for local variables and parameters:
see http://stackoverflow.com/questions/9579403/intellij-idea-automatically-add-final-keyword-to-the-generated-variables
Maybe we should all set that in our IntelliJ and also have a way of sharing such settings (quite easy for intelliJ, we share it also in my current project).

@danielsuter 
In the current PR there have been many finals removed for unknown reasons. That is not soo cool. 
What do you propose for the future concerning that?

**Comment by bruderol** [comment]
@adiherzog Kannst Du ev. mal prüfen, ob mit folgender Anpassung in Eclipse der Formatter die Klasse DummyIdeFormattingSettingTestClass genau gleich formatiert, wie dies intelliJ auf Dani's Branch bereits formatiert hat?

siehe https://github.com/scenarioo/scenarioo/wiki/Eclipse-IDE-Settings/_compare/14d4fef7e336a405d6682d874037184727e0e85f...33b4b7453aa628b3e59849eeafa9bf0613603784

Dann müssten wir das nur bei allen anpassen und mal allen Java-Code umformattieren und dann wären wir mit IntelliJ formattierungen besser aligned.

So oder so sehe ich mit Eclipse wneiger ein Problem, da wir das dort mal so eingestellt haben, dass Eclipse nur Stellen umformattiert, die wirklich auch geändert wurden.
Und so sollten wir das möglichst auch in intelliJ handhaben.

**Comment by bruderol** [comment]
Und hioer noch ein paar weitere Abklärungen zu formattierungs-Einstellungen in IntelliJ

Autoformat on Save für intelliJ (scheint nur schwierig möglich):
http://stackoverflow.com/questions/17879475/how-enable-auto-format-code-for-intellij-idea

Immer nur geänderte Codeblöcke formattieren:
Scheint irgendwie in IntelliJ auch nicht einfach so einstellbar (ausser mit Ctrl+Shift+Alt+L und dann immer `only VCS changed text` auswählen), very annoying! 

Somit muss das wohl jeder bei jedem Commit selber sicherstellen und ich schlage deshalb vor, dass wir das einfach in unsere Developer Guidelines aufnehmen, ungefähr wie folgt:

Bei jedem Commit ist sicherzustellen, dass alle Changes an Files im Commit folgenden Regeln entsprechen:
1. Unsere Coding Guidelines und Code Formatiierungsregeln sind in allen geänderten Code-Abschnitten/Zeilen möglichst eingehalten.
2. Es wurden keine Umformattierungen im File vorgenommen an Stellen im File, die sonst nicht geändert werden mussten (dies führt sonst zu schwierigem Nachvollziehen der wirklichen erfolgten Änderungen und potentiel mehr Merge-Konflikten).
3. Keine unnötige und nicht klar begründbare Verschiebungen / Rearrangements von existierenden Codeblöcken/Funktionen in Source-Files, z.B. auf keinen Fall eine automatische Sortierung von Funktionen (wie z.B. alphabetisch), so etwas entspricht weder unseren Coding Standards noch irgendwelchen üblichen Code Conventions oder Best Practices.
4. Es ist insbesondere auch nicht hilfreich Leerzeilen oder Spaces/Tabs/Indentations an Stellen einzufügen, wo sonst keine Änderung erfolgt ist.
5. Es wurden keine `final` keywords entfernt, die eigentlich richtig sind.
6. Ein commit darf keine "fit" und "fdescribe" für fokusierte Jasmine-Tests enthalten. (für letzteres sollten wir mal einen Git Precommit einrichten)

@adiherzog @danielsuter Weiteres? Was meint ihr? Wollen wir das so machen? Oder seht ihr eine andere Möglichkeit?

**Comment by bruderol** [comment]
@danielsuter Du kannst dann überigens somit die Änderungen in der Klasse DummyIdeFormattingSettingTestClass vom PR so lassen und mergen, wie sie jetzt ist. Ich glaub wir haben das dazumals bewusst so geändert, weil es in IntelliJ nicht anders geht. Richtig?

**Comment by danielsuter** [comment]
@bruderol 
Ich weiss das nicht mehr, kann sehr gut auch ein "Fehler" gewesen sein.

Btw. Ich finde wir sollten unbedingt diskutieren, ob wir überall finals wollen. Ich persönlich finde das extrem störend für den Lesefluss.

**Comment by bruderol** [comment]
betreff final @danielsuter :
1. Wenn man sich mal dran gewöhnt hat, stört es nicht
2. Es gibt schon genügend Diskussionen dazu im Internet, siehe z.B. http://stackoverflow.com/questions/137868/using-final-modifier-whenever-applicable-in-java für einige gute Argumente dafür, warum das empfohlen wird. 
3. Ich persönlich fand es am Anfang genau so störend und unnötig. Irgendwann habe ich sie aber schätzen gelernt und gemerkt, dass es sehr nützlich ist (wenn es konsistent gesetzt wird!). Mittlerweile habe ich mich so daran gewöhnt, dass ich es vermissen werde und mich dann immer fragen muss, ob das jetzt final ist oder eben nicht. Vermutlich werden wir auch den einen oder anderen Bug mehr haben deswegen.
4. Störend finde ich es eigentlich eher, wenn final nur inkonsistent verwendet wird (also nach Lust und Laune, dann bringts mit nämlich beim Lesen wirklich nichts).
5. Falls es aber aufgrund der enforcten `final` in unserem Code zu irgendwelchen Problemen kommt, dann lasse ich mich gerne überzeugen (mit etwas besseren Argumenten als nur "störend für den Lesefluss") und wir können in Zukunft gerne darauf verzichten dies zu enforcen.
6. ABER: wenn es schon final drin hat, dann muss man es ja nicht überall entfernen und dadurch unnötige Changes im Changeset erzeugen, oder???
7. Ihr könnt das gerne diskutieren und entscheiden. Ich beuge mich dann der Mehrheit aller Java-Entwickler in diesem Projekt. Mich würden einfach noch die Argumente interessieren. Danke.

**Comment by danielsuter** [comment]
Naja ich bin wie auch auf Stackoverflow der Meinung des Kommentars mit am meisten Votes. Keine finals bei lokalen Variablen oder Methoden, sonst überall. Aber natürlich werde auch ich mich der Mehrheit beugen.

**Comment by bruderol** [comment]
@danielsuter Es gibt auch eine zweite Antwort (die erste, welche auch vom Fragesteller als beste Antwort akzepotiert wurde) die fast so viele Votes hat und die Argumente darauf solltest du wirklich mal lesen. Und auch die andere Antwort mit den meisten Votes besagt ja nicht dass man finals für Parameter und lokale Variablen verbeiten oder gar nie verwenden soll.
Ich fands gerade für Parameter noch praktisch, weil es häufig passiert dass man den Parameter versehentlich überschreibt, was manchmal zu schwierig zu findenden Bugs führen kann und weil ich der Meinung bin, dass es eigentlich nie einen wirklich guten Grund gibt einen Parameter in einer Methode drin zu ändern. 
Ausserdem ist es praktisch wenn sie schon final sind, falls man sie in einem inneren Scope verwenden muss (Listener oder so).
Ich bin da aber nicht dogmatisch und habe kein Problem wenn wir die finals im code für Parameter nicht immer 100% sicherstellen können.
Deshalb ja mein Vorschlag: 
1. Wer eine Methode scheibt, soll selber entscheiden dürfen, ob er final auf Parametern und lokalen Variablen verwenden will (da bin ich echt schmerzfrei).
2. ABER UND DAS IST WICHTIG: Wer Code von anderen anpasst, entfernt nicht unnötigerweise finals.

Also: Was spricht jetzt da dagegen oder was ist jetzt da das Problem damit bitteschön?
Let's discuss it wenn wir uns das nächste mal treffen.
Danke!

**Comment by bruderol** [comment]
Wegen den Warnings, die @adiherzog noch gefunden hat:
Diese werden in intelliJ schon angezeigt, sie wurden aber wohl ignoriert, weil es etwa 1000 andere Warnungen gibt. Wir müssen da die Einstellungen noch anpassen, dass Warnungen die für uns nicht relevant sind, nicht mehr erscheinen. Ein Beipsiel: IntelliJ gibt eine Warnung aus, für nicht explizt behandelte Runtime-Exceptions. Komisch.

**Comment by bruderol** [comment]
@adiherzog @danielsuter finde wir sollten dieses Thema mal abschliessen - gibt es da noch was zu tun? Ev. mal allen Java-Code umformattieren bevor wir im 2018 gröbere Server-Umbauten beginnen und noch mehr Leute gleichzeitig dran arbeiten??

**Comment by bruderol** [comment]
@adiherzog @danielsuter I opened a new issue #655 concerning the code formatting topic.

Let us quickly verify whether everything else is ok now with the IntelliJ setup and then do the following to close this issue:

1. cleanup docu for the dev env: such that we have on only truth and this is the IntelliJ setup, explained in such a way, that you are free to use other IDE, but mostly targeted to devs using IntelliJ which is our primary choice
   * remove old docu parts not needed anymore
   * consolidtae parts that are currently duplicated
2. open other issues if there is some further left overs (?)
3. close this issue

**Comment by bruderol** [comment]
Thanks @cgrossde - yes we can still open new issues if we find leftovers that need to be correctd. Thanks
